### ISSUE TITLE
Limiting concurrent sessions

### ISSUE THREAD

**Issue by thelonious89** [Jigsaw-Code/outline-server#439]
Is there a built-in fix yet for limiting concurrent sessions? Similar to this request here:
https://github.com/Jigsaw-Code/outline-server/issues/32#issuecomment-379418574

It’s important as one part of an access protocol in a security regime.

**Comment by fortuna** [comment]
Do you mean the number of different devices using an access key?
Could you clarify why you need that?

**Comment by thelonious89** [comment]
Yes, the number of different devices using an access key at the same time.
It's important because:
1) if the confidentiality of the keys are broken, then at least we can limit access. So if I give employee Bob the key to put on his company device, he cannot share it with Alice's device.
2) Track the bandwidth and usage better, so I know it's all coming from 1 device rather than maybe an entire office of machines.

Thank you~

**Comment by fortuna** [comment]
Thanks for the feedback. We are looking into implementing features to allow admins to control access to their servers.

We are currently focused on implementing quotas, but we'll consider this after that.

**Comment by thelonious89** [comment]
Excellent thank you. Yes quotas are actually the best ultimate way to stop bandwidth abuse because even one user can fry the lines. So working on quotas should be the first priority.

**Comment by remagio** [comment]
Any updates about this features request...?

**Comment by HosseyNJF** [comment]
This feature is really needed...

**Comment by fortuna** [comment]
FYI, data limits has launched already. We may tweak it a bit.
W are focused on server resilience, making sharing easier and adding more cloud providers.
Limiting concurrent connections is something we've thought about, but it's not in the immediate roadmap.

**Comment by HosseyNJF** [comment]
@fortuna I like to contribute to this feature because I need it personally. could you give me a little guidance so that I can start developing the feature?

**Comment by halpro** [comment]
Hello team, this is really annoying that the entire world is asking you to implement this single essential feature for YEARS BUT you continue to polish the UI.

**Comment by crash481** [comment]
@halpro we can only ask developers to do it. Show respect and gratitude to the developers for making it OpenSource or contribute and make it yourself

Also I vote that this will be a good feature

**Comment by crash481** [comment]
@HosseyNJF did you contributed and have some success on implement it?

**Comment by HosseyNJF** [comment]
@crash481 I did try, but it is impractical at best. The best method is to limit by IP count (prevent more than 2 IPs from using the service simultaneously), but I don't have the resources and time to develop it. :(

**Comment by BolshakovEvgeniy** [comment]
> @crash481 I did try, but it is impractical at best. The best method is to limit by IP count (prevent more than 2 IPs from using the service simultaneously), but I don't have the resources and time to develop it. :(

IP count isn't best method, cause different devices with same key can be hidden under NAT, for example. 

The best method, in my opinion, is to simply add an incremental counter (or even a boolean flag) and break the connection if it's set to some value. Thus, the first connected user sets the flag to the "busy" state, the second and subsequent ones receive a connection break.
